### PR TITLE
Add support to explicitly add priority for the handshakers using enums

### PR DESCRIPTION
**Pull request by anicr7**
Currently the order of the handshakers is controlled by a single bool(at_start). This doesn't allow for more complex use cases where the handshaker has to be done before tcp connect handshaker for example.

By explicitly adding enums that specify the priority, we allow for a cleaner abstraction for registering handshakers.

### CODE REVIEW

File: src/core/lib/transport/handshaker_factory.h [hunk 1]
```diff
@@ -35,11 +37,38 @@ namespace grpc_core {
 
 class HandshakeManager;
 
+// Enum representing the priority of the client handshakers.
+// The order of the client handshakers is decided by the priority.
+// For example kPreTCPConnect handshakers are called before kTCPConnect and so
+// on.
+enum class HandshakerClientPriority : int {
+  kPreTCPConnect,
```
Comment (ctiller) [inline src/core/lib/transport/handshaker_factory.h:45]: Probably worth expanding the name and including one line on the kinds of things expected in each bucket

File: src/core/lib/transport/handshaker_factory.h [hunk 2]
```diff
@@ -35,11 +37,38 @@ namespace grpc_core {
...
+  kHTTPConnect,
+  kSecurity,
+};
+
+// Enum representing the priority of the server handshakers.
+// The order of the server handshakers is decided by the priority.
+// For example kReadHeadSecurity handshakers are called before kSecurity and so
+// on.
+enum class HandshakerServerPriority : int {
+  kReadHeadSecurity,
```
Comment (ctiller) [inline src/core/lib/transport/handshaker_factory.h:56]: here too

File: src/core/lib/transport/handshaker_factory.h [hunk 3]
```diff
@@ -35,11 +37,38 @@ namespace grpc_core {
...
+  kSecurity,
+};
+
+using HandshakerPriority =
```
Comment (ctiller) [inline src/core/lib/transport/handshaker_factory.h:60]: This is probably reaching for a too large a hammer for the job.

Consider folding HandshakerClientPriority/HandshakerServerPriority into one enum ... we can comment client/server only if necessary (heck, we can introduce a IsServer/IsClient if necessary too... but I don't think it's practically useful)

**Review by ctiller — CHANGES_REQUESTED** [review]

File: src/core/lib/transport/handshaker_factory.h (hunk 2 above)
Comment (anicr7) [reply src/core/lib/transport/handshaker_factory.h]: Done.

File: src/core/lib/transport/handshaker_factory.h (hunk 1 above)
Comment (anicr7) [reply src/core/lib/transport/handshaker_factory.h]: Done, let me know if this good enough or if I should add more information.

File: src/core/lib/transport/handshaker_factory.h (hunk 3 above)
Comment (anicr7) [reply src/core/lib/transport/handshaker_factory.h]: I initially had a single priority for both clients and servers but I thought it looked odd to have PreTCPConnectHandshaker for Server side handshakers. Adding a comment sounds good to me considering the current limited usage.

**Review by ctiller — APPROVED** [review]

File: src/core/lib/transport/handshaker_factory.h [hunk 4]
```diff
@@ -35,11 +35,37 @@ namespace grpc_core {
...
+};
+
 class HandshakerFactory {
  public:
   virtual void AddHandshakers(const ChannelArgs& args,
                               grpc_pollset_set* interested_parties,
                               HandshakeManager* handshake_mgr) = 0;
+  // Return the priority associated with the handshaker.
+  // The priority can be either HandshakerClientPriority or
+  // HandshakerServerPriority depending on the type of handshaker.
```
Comment (markdroth) [inline src/core/lib/transport/handshaker_factory.h:67]: This comment looks out of date.  There is no `HandshakerServerPriority`.

File: src/core/lib/transport/handshaker_factory.h [hunk 5]
```diff
@@ -35,11 +35,37 @@ namespace grpc_core {
 
 class HandshakeManager;
 
+// Enum representing the priority of the handshakers.
+// The order of the handshakers is decided by the priority.
+// For example kPreTCPConnect handshakers are called before kTCPConnect and so
+// on.
+enum class HandshakerPriority : int {
```
Comment (markdroth) [inline src/core/lib/transport/handshaker_factory.h:42]: Suggest moving this inside of the `HandshakerFactory` class.

File: src/core/lib/transport/handshaker_factory.h [hunk 6]
```diff
@@ -35,11 +35,37 @@ namespace grpc_core {
...
+  kPreTCPConnectHandshakers,
+  // Handshakers responsible for the actual TCP connect establishment.
+  // Applicable mainly for Client handshakers.
+  kTCPConnectHandshakers,
+  // Handshakers responsible for the actual HTTP connect established. Applicable
+  // mainly for Client handshakers.
+  kHTTPConnectHandshakers,
+  // Handshakers that should be called before security handshakes but after
+  // connect establishment. Applicable mainly for Server handshakers currently.
+  kReadHeadSecurityHandshakers,
```
Comment (markdroth) [inline src/core/lib/transport/handshaker_factory.h:54]: This should be called `kReadAheadSecurityHandshakers` ("Ahead" instead of "Head").

**Review by markdroth — APPROVED** [review]
This looks good overall.  Comments are all cosmetic.

File: src/core/lib/transport/handshaker_factory.h (hunk 4 above)
Comment (anicr7) [reply src/core/lib/transport/handshaker_factory.h]: Done, removed.

File: src/core/lib/transport/handshaker_factory.h (hunk 5 above)
Comment (anicr7) [reply src/core/lib/transport/handshaker_factory.h]: Done.

File: src/core/lib/transport/handshaker_factory.h (hunk 6 above)
Comment (anicr7) [reply src/core/lib/transport/handshaker_factory.h]: Done.

**Review by markdroth — APPROVED** [review]

**Pull request merged**